File: BFS_DFS_etc_Data_structure/all_of_graph/graph-8way.py
"""
Given a 2D grid with '1' and '0', two '1' cells belong to the same island if 
they are connected in any of the 8 directions (up, down, left, right, 
and the 4 diagonals). Count the total number of islands under this rule.
"""
from typing import List

def count_islands_8dir(M: List[List[str]]) -> int:
    if not M or not M[0]:
        return 0
    rows, cols = len(M), len(M[0])
    visited = [[False]*cols for _ in range(rows)]

    # 8 directions
    directions = [(-1,0),(1,0),(0,-1),(0,1),
                  (-1,-1),(-1,1),(1,-1),(1,1)]
    
    def dfs(r, c):
        stack = [(r,c)]
        visited[r][c] = True
        while stack:
            rr, cc = stack.pop()
            for dr, dc in directions:
                nr, nc = rr + dr, cc + dc
                if 0 <= nr < rows and 0 <= nc < cols:
                    if M[nr][nc] == '1' and not visited[nr][nc]:
                        visited[nr][nc] = True
                        stack.append((nr, nc))
    
    count = 0
    for r in range(rows):
        for c in range(cols):
            if M[r][c] == '1' and not visited[r][c]:
                dfs(r, c)
                count += 1
    return count

if __name__ == "__main__":
    grid = [
        ['1','1','0','0'],
        ['0','1','0','1'],
        ['1','0','1','1'],
        ['0','0','0','0']
    ]
    # The exact island count depends on diagonal connectivity.
    ans = count_islands_8dir(grid)
    print("Number of Islands (8-dir):", ans)
    # No explicit 'expected' here because it depends on how the cells connect diagonally.

File: BFS_DFS_etc_Data_structure/all_of_graph/graph-8way.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int R, C;
char** Map;
bool** visited;

int dr8[8] = {-1, 1, 0, 0, -1, -1, 1, 1};
int dc8[8] = {0, 0, -1, 1, -1, 1, -1, 1};

typedef struct {
    int r, c;
} StackItem;

void dfs_8dir_stack(int sr, int sc){
    StackItem* stack = (StackItem*)malloc(R*C*sizeof(StackItem));
    int top = -1;
    
    top++;
    stack[top].r = sr;
    stack[top].c = sc;
    visited[sr][sc] = true;
    
    while(top >= 0){
        StackItem cur = stack[top--];
        int rr = cur.r, cc = cur.c;
        
        for(int i=0; i<8; i++){
            int nr = rr + dr8[i];
            int nc = cc + dc8[i];
            if(nr>=0 && nr<R && nc>=0 && nc<C){
                if(Map[nr][nc] == '1' && !visited[nr][nc]){
                    visited[nr][nc] = true;
                    top++;
                    stack[top].r = nr;
                    stack[top].c = nc;
                }
            }
        }
    }
    free(stack);
}

int count_islands_8dir(char** grid, int rows, int cols){
    R = rows; C = cols;
    Map = grid;
    
    // visited 2D
    visited = (bool**)malloc(R*sizeof(bool*));
    for(int i=0; i<R; i++){
        visited[i] = (bool*)calloc(C, sizeof(bool));
    }
    
    int island_count = 0;
    for(int r=0; r<R; r++){
        for(int c=0; c<C; c++){
            if(Map[r][c]=='1' && !visited[r][c]){
                dfs_8dir_stack(r, c);
                island_count++;
            }
        }
    }
    
    // cleanup
    for(int i=0; i<R; i++){
        free(visited[i]);
    }
    free(visited);
    return island_count;
}

int main(void){
    int rows = 4, cols = 4;
    char* sample[4] = {
        "1100",
        "0101",
        "1011",
        "0000"
    };
    
    char** grid = (char**)malloc(rows*sizeof(char*));
    for(int i=0; i<rows; i++){
        grid[i] = sample[i];
    }
    
    int ans = count_islands_8dir(grid, rows, cols);
    printf("Number of Islands (8-dir): %d\n", ans);
    
    free(grid);
    return 0;
}
